### DISCUSSION TITLE
ESP32 timing issues when sending a raw code

### DISCUSSION THREAD

**Discussion by cssyria** [Arduino-IRremote/Arduino-IRremote#1138] [Help]
I'm using Arduino-IRremote in an ESP32-based project with good success. Recently, we wanted to add some additional IR commands from an Xfinity cable box remote. I captured the remote using a Global Cache iTach device, and got the following timings:
`+210,-823,+210,-1758,+210,-823,+210,-2703,+210,-1293,+210,-1293,+210,-1293,+210,-2703,+210,-12989,+210,-823,+210,-2703,+210,-823,+210,-823,+210,-823,+210,-823,+210,-823,+210,-823,+210,-82369,+210,-823,+210,-1758,+210,-823,+210,-2703,+210,-1293,+210,-1293,+210,-1293,+210,-2703,+210,-12989,+210,-823,+210,-1293,+210,-1758,+210,-823,+210,-823,+210,-823,+210,-823,+210,-823,+210`

When I put these into my code and send them using sendRaw:
```
    const uint16_t irSignal[] = { 210, 823, 210, 1758, 210, 823, 210, 2703, 210, 1293, 210, 1293, 210, 1293, 210, 2703, 210, 12989, 210, 823, 210, 2703, 210, 823, 210, 823, 210, 823, 210, 823, 210, 823, 210, 823, 210, 82369, 210, 823, 210, 1758, 210, 823, 210, 2703, 210, 1293, 210, 1293, 210, 1293, 210, 2703, 210, 12989, 210, 823, 210, 1293, 210, 1758, 210, 823, 210, 823, 210, 823, 210, 823, 210, 823, 210 }; // Comcast X1 Menu key
    IrSender.sendRaw(irSignal, sizeof(irSignal) / sizeof(irSignal[0]), STD_KHZ); // Note the approach used to automatically calculate the size of the array.
    delay(1000); // delay must be greater than 5 ms (RECORD_GAP_MICROS), otherwise the receiver sees it as one long signal
```

I'm getting some discrepancy with the timing, as measured on the iTach receiver:
`+211,-827,+211,-1440,+211,-827,+211,-2711,+211,-1440,+211,-1440,+211,-1440,+211,-2711,+211,-13002,+211,-827,+211,-2711,+211,-827,+211,-827,+211,-827,+211,-827,+211,-827,+211,-827,+211,-82357,+211,-827,+211,-1440,+211,-827,+211,-2711,+211,-1440,+211,-1440,+211,-1440,+211,-2711,+211,-13002,+211,-827,+211,-1440,+211,-1440,+211,-827,+211,-827,+211,-827,+211,-827,+211,-827,+211`
Essentially, my 1750us and 1300us spaces are all coming out at 1440us, leaving the IR code undetected by the receiving cable box. The other marks and spaces all seem to be timed as expected.

I've tested this with the SendRawDemo example. The NEC code originally included with the example transmits as expected, with the correct timing at the iTach receiver. Modifying the array of timings with mine gives the same inaccuracy on the 1750 and 1300 spaces. I also tried rounding every value to the nearest 50us, with no change.

I'm not sure why this is occurring on my codes. We've had excellent luck sending NEC/RC5/RC6 codes and even other raw codes in the past. Our project was originally built using a fork from the 2.2.3 branch, but this issue was experienced on 2.8 and 4.1.2 during testing.

**Comment by cssyria** [comment]
Additional datapoint - the same issue seems to occur when using sendPronto as well. 

Using this test code modified from SendProntoDemo and the Pronto hex output from IrScrutinizer:
```
    IrSender.sendPronto("0000 006D 0024 0000 0008 001F 0008 0043 0008 001F 0008 0067 0008 0031 0008 0031 0008 0031 0008 0067 0008 01EF 0008 001F 0008 0067 0008 001F 0008 001F 0008 001F 0008 001F 0008 001F 0008 001F 0008 0C43 0008 001F 0008 0043 0008 001F 0008 0067 0008 0031 0008 0031 0008 0031 0008 0067 0008 01EF 0008 001F 0008 0031 0008 0043 0008 001F 0008 001F 0008 001F 0008 001F 0008 001F 0008 130E" 
    , 0); // No repeat possible, because of missing repeat pattern
```

Captured output:
```
+212,-816,+212,-1429,+212,-816,+212,-2685,+212,-1429,+212,-1429,+212,-1429,+212,-2685,+212,-12871,+212,-816,+212,-2685,+212,-816,+212,-816,+212,-816,+212,-816,+212,-816,+212,-816,+212,-65644,+212,-816,+212,-1429,+212,-816,+212,-2685,+212,-1429,+212,-1429,+212,-1429,+212,-2685,+212,-12871,+212,-816,+212,-1429,+212,-1429,+212,-816,+212,-816,+212,-816,+212,-816,+212,-816,+212,-127997
```

**Comment by AnalysIR** [comment]
This appears to XMP, which is very difficult to capture unless you have a receiver designed for XMP. IR receivers can distort timings by up to +/- 200 uSecs....so with the nature of XMP it can be difficult to get right.

Using pronto also makes things worse as there is a built in inaccuracy in the design of this encoding. So you will suffer from a double whammy for XMP + PRONTO + IR receiver.

try this....it will depend on how good your original capture was.

/*
Automatically Generated by AnalysIR - Batch Export Utility
Registered to: info@analysir.com
Session History
Type : Key : Value : Bits : Carrier Frequency (kHz)
0 : XMP :  : 070E444E0E000000070E444E04700000 : 128 : 0
Note: Be sure to use the correct Carrier frequency, for each individual signal, as(or if) indicated above
*/

// NB: Not all protocols are supported by IRremote or IRLib. You may need to edit the code below manually
// Automatically Generated by AnalysIR for info@analysir.com, visit https://www.AnalysIR.com/ or email info@....... for further details
int khz=38; //NB Change this default value as necessary to the correct carrier frequency


uint16_t Signal_0_0[] = {212,758,212,1715,212,758,212,2673,212,1305,212,1305,212,1305,212,2673,212,12989,212,758,212,2673,212,758,212,758,212,758,212,758,212,758,212,758,212,82369,212,758,212,1715,212,758,212,2673,212,1305,212,1305,212,1305,212,2673,212,12989,212,758,212,1305,212,1715,212,758,212,758,212,758,212,758,212,758,212}; //AnalysIR Batch Export (IRremote) - RAW

irsend.sendRaw(Signal_0_0, sizeof(Signal_0_0)/sizeof(int), khz); //AnalysIR Batch Export (IRremote) - RAW
 // AnalysIR IR Protocol: XMP, Key:
